Npfreview -  ALTQ integration in NPF

Emmankoko opened the pull request:
This is a PR opened solely for review purpose. after all checks are cleared, it will be merged into the CVS main source.
@zoulasc @MartinHusemann

Emmankoko committed:
Compile-time error fix in altq

use right jobs altq interface to be detached from and set your error variable

Emmankoko committed:
Plain grammar for ALTQ configuration in /etc/npf.conf

this follows a style.....:

altq on $interface $your_queue_options $your_queue_managements {child queues here}
then....
queue child_queue and its defintions

example for cbq:
altq on wm0 cbq qlimit 60 tbrsize .. bandwidth 10Mb queue {deflt, audio, video}
queue deflt bandwidth 1Mb priority 1 cbq(default, red)
queue audio bandwidth 2Mb priority 2 cbq(borrow, ecn)
queue video bandwidth 3Mb priority 3 cbq(red)

example for hfsc:
altq on wm0 hfsc qlimit 60 tbrsize .. bandwidth 10Mb queue {deflt, audio, video}

queue deflt bandwidth 1Mb priority 1 hfsc(default, red)
queue audio bandwidth 2Mb priority 2 hfsc( ecn)
queue video bandwidth 3Mb priority 3 hfsc(red)

for CBQ and HFSC, CBQ borrows unlike HFSC.
example for priq:

altq on wm0 qlimit 60 tbrsize .. bandwidth 10Mb queue { audio, video }
queue audio priority 1 priq(default)
queue video priority 2 priq(ecn)

NB in PRIQ: you cannot set bandwidth for child queues in PRIQ as child queues depend on priority
NB in ALL: all parents must have default queueus

on a rule....

can be appended on a rule as...
pass in on wm0 from $mac to $bsd apply "log" queue "audio".

Emmankoko committed:
Integrate NPF into ALTQ: might be a goodbye for PF & ALTQ

when altq is being attached to a network interface through a packet filter(PF), it uses pf_altq
which is a structure containing all pf's elements necessary for use in adding queues, removing queues,

similar structure is implemented for npf (struct npf_altq) and it's being used to substitute pf structure.

this automatically kills pf in altq.

this hits out backs to move npf to where we want to take it to.

if we want pf still supported in altq,

maybe a
functions with (pf_altq)

same functions with (npf_altq)

going to make code messy and unattractive.
but I'm ready to listen to your suggestions.

Emmankoko committed:
kernel component that loads altq configurations into altq framework

three IOCLS:

IOC_NPF_BEGIN_ALTQ:
acts as a safe guard to only initialize and allocate memory when adding queue
for the first time or adding after a flush. if altq is already loaded, it doesn't initlise
or allocate, just clears already existing altq and ensures availablity of ifq structure to safely take on new queues.

IOC_NPF_ADD_ALTQ:
adds queues to the altq framework(using altq_add) which is called directly into ALTQ.
queues are represented with npf_altq structure linked together in a TAILQ.(categorized into active and inactive).

IOC_NPF_GET_ALTQ:
checks if altq is supported to allow the queues to be sent to the kernel.

Emmankoko committed:
Userland code parses the altq configurations and sends to kernel

main functions are :
expand_altq and expand_queue.

expand_altq acts on:
altq on $interface $scheduler $queue_options $child_queues

expand_queue acts on:
defined child queues..
queue blabla bandwidth x priotiy y ....

the rest for calcualting bandwidth, tbrsizes, qlenght etc.

most were re-implemeted from PF and has been well tested.

safety guards:

before a succesful addition into the kernel, it's checked if only on discipline is being loaded
as it doesn't allow the defintion of two or more disciplines.

Emmankoko committed:
userland code to append queue to rule

in the userland, it's main job is to enure the queue is defined and also the altq config
conforms with the best practices. i.e: a default rule must be present and cbq and hfcs must ahve root queues

afterwords, sets nvlist name of "queue" and value of "the queue name" and appends it to the rule nvlist"

userland npf_altq.c has been renamed to npfctl_altq.c

Emmankoko committed:
kernel code to append queue to a rule

basically, when a queue is found, on a rule, the queueing discipline of that queue is attached
to the network interface using the npf_altq interface. done in the npf_altq_commit function.
this sets the enqueue, dequeue, request, disc type etc. on the network interface.

and then altq is enabled on the network interface to allow for queueing.

when a packet is passed by the filter, it is then tagged with an ALTQ tag with the qid of the queue being set on the rule and then enqueued in the kernel

Emmankoko committed:
flushing config with altq

this is the flush implementation which involes;

npfctl flush: which flush both altq and filtering

if altq isn't loaded, flushes only filtering

or

npfctl flush -q: to flush only altq
NB:

	if (npf_altq_loaded) {
		pool_destroy(&npf_altq_pl);
		npf_altq_loaded = 0;
	}
serves as safe guard to avoid calling pool_destroy twice or more if a user runs:
npfctl flush
npfctl flush.

in a first flush, npf_altq_loaded is set to 0 so a second nffctl flush skips the block

Emmankoko committed:
starting ALTQ in npf

there's no special command required to start ALTQ.

when you
npfctl start

it checks if any altq configuration is added in the config and then starts altq when npf is starting

if no queue config is present, it ignores altq start

Emmankoko committed:
stop queueing

stopping queueing also requires no special commands.

when a user
npfctl stop:

it checks if queuing is active, if it is, it stops it.

if altq isn't running, it ignores it and detach the filtering hooks

Emmankoko committed:
print statistics on queueing and filtering

"npfctl stats" have been modified into

npfctl stats -f | -q

nfctl stats -q prints queueing statistics and npfctl stats -f prints filtering statistics.

with npfctl stats -q,

it calls into the kernel npf_altq_get_qstats which then calls altq_getstats on the particular queue.

then drops, xmit counts, queued counts etc are printed out.

Emmankoko committed:
show altq using npfctl show -q

showing reloaded queues is triggered passing a -q flag on npfctl show.

npfctl show displays loaded filtering rules and also displays the state of altq in npf(either enabled or disabled).

zoulasc commented on sys/altq/altq_jobs.c:
```diff
@@ -1831,14 +1831,14 @@ int
 jobsclose(dev_t dev, int flag, int fmt,
     struct lwp *l)
 {
+	int error = 0;
```
Revert this, and return 0 at the end.

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -181,8 +203,9 @@ yyerror(const char *fmt, ...)
 %token	<str>		PARAM
 %token	<str>		TABLE_ID
 %token	<str>		VAR_ID
+%token  <str>       BW_SPEC
```
whitespace

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -199,6 +222,15 @@ yyerror(const char *fmt, ...)
 %type	<filtopts>	filt_opts all_or_filt_opts
 %type	<optproto>	rawproto
 %type	<rulegroup>	group_opts
+%type	<queue_opts> queue_opts queue_opt queue_opts_l
```
whitespace

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -199,6 +222,15 @@ yyerror(const char *fmt, ...)
+%type	<queue_options>	scheduler
+%type	<num>		cbqflags_list cbqflags_item
+%type	<num>		priqflags_list priqflags_item
+%type	<hfsc_opts>		hfscopts_list hfscopts_item hfsc_opts
```
whitespace

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -199,6 +222,15 @@ yyerror(const char *fmt, ...)
+%type	<num>		priqflags_list priqflags_item
+%type	<hfsc_opts>		hfscopts_list hfscopts_item hfsc_opts
+%type	<queue_bwspec>	bandwidth
+%type   <str>			queue_flags
```
whitespace

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -232,9 +270,346 @@ line
+altq : ALTQ on_ifname queue_opts QUEUE qassign {
+		struct npf_altq a;
+
+		memset(&a, 0, sizeof(a));
```
why do we need memset? which fields aren't we initializing?

zoulasc replied:
ah a.qname. Ok, move the initialization after the first yyerror though.

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -232,9 +270,346 @@ line
+		;
+
+queue_opts	:	{
+			memset(&queue_opts, 0, sizeof queue_opts );
```
sizeof(queue_opts)

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -232,9 +270,346 @@ line
+		    queue_opts_l
+			{ $$ = queue_opts; }
+		| /* empty */ {
+			memset(&queue_opts, 0, sizeof queue_opts);
```
again parens in sizeof

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -232,9 +270,346 @@ line
+		| queue_opt
+		;
+
+queue_opt	: BANDWIDTH bandwidth	{
```
whitespace before braces everywhere should be consistent.

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -232,9 +270,346 @@ line
+bandwidth	: bw_spec {
+			struct node_queue_bw bw;
+
+			if (npfctl_eval_bw(&bw, $1)){
```
space before brace

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -232,9 +270,346 @@ line
+		;
+
+bw_spec : BW_SPEC
+		{$$ = $1; }
```
whitespace.

zoulasc commented on usr.sbin/npf/npfctl/npf_parse.y:
```diff
@@ -232,9 +270,346 @@ line
+		;
+
+qassign_item	: IDENTIFIER			{
+			$$ = calloc(1, sizeof(struct node_queue));
```
doesn't sizeof(*$$) work?

zoulasc commented on usr.sbin/npf/npfctl/npf_scan.l:
```diff
@@ -174,6 +175,19 @@ icmp-type		return ICMPTYPE;
 any			return ANY;
 
+altq			return ALTQ;
+cbq				return CBQ;
```
whitespace

zoulasc commented on usr.sbin/npf/npfctl/npfctl.h:
```diff
@@ -48,6 +48,69 @@
+struct node_queue_opt {
+	int			 qtype;
+	union {
+		struct npf_cbq_opts		cbq_opts;
```
whitespace?

zoulasc commented on sys/altq/altq_hfsc.c:
```diff
@@ -51,6 +51,11 @@ __KERNEL_RCSID(0, "$NetBSD: altq_hfsc.c,v 1.30 2021/09/21 14:30:15 christos Exp
 #include "pf.h"
 #endif
 
+#ifndef NPF
```
why is this needed?

zoulasc commented on sys/net/npf/npf_altq.h:
```diff
@@ -0,0 +1,102 @@
+/*-
```
/* $NetBSD$ */

Emmankoko replied:
`/*	$NetBSD: apm.c,v 1.35 2021/09/26 01:16:08 thorpej Exp $ */`
are you talking about something like this ?? thought that will be done when moving into CVS ??

zoulasc commented on sys/net/npf/npf_altq.h:
```diff
@@ -0,0 +1,102 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <sys/queue.h>
```
Why outside the ifdef?

zoulasc commented on sys/net/npf/npf_altq.h:
```diff
@@ -0,0 +1,102 @@
+ * options defined on the cbq, priq and hfsc when configuring them
+ */
+struct npf_cbq_opts {
+	u_int		minburst;
```
I would rather use sized types (uint32_t ).

zoulasc commented on distrib/sets/lists/comp/mi:
```diff
@@ -2762,6 +2762,7 @@
 ./usr/include/net/net_stats.h			comp-c-include
 ./usr/include/net/netisr.h			comp-obsolete		obsolete
 ./usr/include/net/npf.h				comp-c-include
+./usr/include/net/npf_altq.h        comp-c-include
```
whitespace.

zoulasc commented on sys/net/npf/npf.h:
```diff
@@ -322,6 +331,9 @@ typedef struct npf_ioctl_table {
 #define	IOC_NPF_CONN_LOOKUP	_IOWR('N', 108, nvlist_ref_t)
 #define	IOC_NPF_TABLE_REPLACE	_IOWR('N', 109, nvlist_ref_t)
+#define IOC_NPF_ADD_ALTQ	_IOWR('N', 110, struct npfioc_altq)
+#define IOC_NPF_GET_ALTQS	_IOWR('N', 111, struct npfioc_altq)
```
Why plural here?

zoulasc commented on sys/net/npf/npf_altq.c:
```diff
@@ -0,0 +1,162 @@
+ * All rights reserved.
+ *
+ * This material is based upon work partially supported by The
+ * NetBSD Foundation under a contract with Mindaugas Rasiukevicius.
```
Why his copyright? Didn't you write the code?

zoulasc commented on sys/net/npf/npf_altq.c:
```diff
@@ -0,0 +1,162 @@
+/*-
```
/* $NetBSD$ */

zoulasc commented on sys/net/npf/npf_altq.c:
```diff
@@ -0,0 +1,162 @@
+ * starting point of altq kernel routines
+ */
+bool	 npf_altq_running = false;
+u_int32_t		 nticket_altqs_active;
```
uintX_t everywhere... not u_intX_t

zoulasc commented on sys/net/npf/npf_altq.c:
```diff
@@ -0,0 +1,162 @@
+			npf_qid_unref(altq->qid);
+		pool_put(&npf_altq_pl, altq);
+	}
+	if (error)
```
just return error here, why check? And I am wondering if we need to break if there is an error (i.e. return in the loop).

zoulasc commented on sys/net/npf/npf_altq.c:
```diff
@@ -0,0 +1,162 @@
+    TAILQ_FOREACH(altq, npf_altqs_active, entries)
+        paa->nr++;
+    paa->ticket = nticket_altqs_active;
+	return 0 ;
```
wrong whitespace

zoulasc commented on sys/net/npf/npf_altq.c:
```diff
@@ -0,0 +1,162 @@
+		return error;
+	}
+	TAILQ_INSERT_TAIL(npf_altqs_inactive, altq, entries);
+	memcpy(&paa->altq, altq, sizeof(struct npf_altq));
```
sizeof(paa->altq); use sizeof(var) not sizeof(type) where possible.

zoulasc commented on sys/net/npf/npf_os.c:
```diff
@@ -274,6 +274,16 @@ npf_dev_ioctl(dev_t dev, u_long cmd, void *data, int flag, lwp_t *l)
+	case IOC_NPF_GET_ALTQS:
+		return npf_get_altqs(data);
+	case IOC_NPF_BEGIN_ALTQ:
+	/* initialize all queueing components on the first attempt*/
```
whitespace.

zoulasc commented on usr.sbin/npf/npftest/Makefile:
```diff
@@ -18,7 +18,7 @@ DPADD+=		${LIBNPFTEST}/libnpftest.a
 
 LDADD+=		-lrump -lrumpvfs_nofifofs -lrumpvfs -lrumpuser
-LDADD+=		-lrumpnet -lrumpnet_net	-lrumpdev_bpf
+LDADD+=		-lrumpnet -lrumpnet_net	-lrumpdev_bpf  -lrumpnet_altq
```
whitespace

zoulasc commented on usr.sbin/npf/npfctl/npf_altq.c:
```diff
@@ -0,0 +1,1395 @@
+/*-
```
/* $NetBSD$ */
Date should be 2024 since this is new and your name (in all copyrights)

zoulasc commented on usr.sbin/npf/npfctl/npf_altq.c:
```diff
@@ -0,0 +1,1395 @@
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/queue.h>
+#ifdef __NetBSD__
```
Why NetBSD? Are all those headers needed? " headers should be after < headers

zoulasc commented on usr.sbin/npf/npfctl/npf_altq.c:
```diff
@@ -0,0 +1,1395 @@
+		if (r == NULL) { \
+			r = calloc(1, sizeof(T)); \
+			if (r == NULL) \
+				err(1, "LOOP: calloc"); \
```
err(EXIT_FAILURE everywhere

zoulasc commented on usr.sbin/npf/npfctl/npf_altq.c:
```diff
@@ -0,0 +1,1395 @@
+	do { \
+		T *n; \
+		if (r == NULL) { \
+			r = calloc(1, sizeof(T)); \
```
sizeof(*r) instead of sizeof(T)

zoulasc commented on usr.sbin/npf/npfctl/npf_altq.c:
```diff
@@ -0,0 +1,1395 @@
+npfctl_test_altqsupport(int dev)
+{
+	struct npfioc_altq pa;
+	if (ioctl(dev, IOC_NPF_GET_ALTQS, &pa)) {
```
== -1

Emmankoko replied:
if a different errno is returned, will -1 also be returned ?

zoulasc commented on usr.sbin/npf/npfctl/npf_altq.c:
```diff
@@ -0,0 +1,1395 @@
+	struct npfioc_altq pa;
+	if (ioctl(dev, IOC_NPF_GET_ALTQS, &pa)) {
+		if (errno == ENODEV) {
+			fprintf(stderr, "No ALTQ support in kernel\n"
```
warnx()

zoulasc commented on usr.sbin/npf/npfctl/npf_altq.c:
```diff
@@ -0,0 +1,1395 @@
+}
+/* evaluate bandwidth */
+int
+npfctl_eval_bw(struct node_queue_bw * bw, char * bw_spec)
```
whitespace

zoulasc commented on usr.sbin/npf/npfctl/npf_altq.c:
```diff
@@ -0,0 +1,1395 @@
+			return (-1);
+		}
+	}
+	bw->bw_absolute = (u_int32_t)bps;
```
I would do:
if (cp == NULL) {
    bw->bw_absolute = (uint32_t)bps;
    return 0;
} 
and then handle the non NULL case.

Emmankoko replied:
if you enter just 10 for bandwidth, you hit a syntax error because of the specified rule for bandwidth (must contain a figure and a string(being byte wise or percentage). a NULL cp(string) won't even get evaluated.
 
 checking for a NULL case in the evaluation looks like a code which will be dead.

zoulasc commented on usr.sbin/npf/npfctl/npf_altq.c:
```diff
@@ -0,0 +1,1395 @@
+expand_altq(struct npf_altq *a, const char *ifname,
+    struct node_queue *nqueues, struct node_queue_bw bwspec,
+    struct node_queue_opt *opts)
+{
```
that method is too big, split it.

zoulasc commented on usr.sbin/npf/npfctl/npf_altq.c:
```diff
@@ -0,0 +1,1395 @@
+expand_queue(struct npf_altq *a, const char *ifname,
+    struct node_queue *nqueues, struct node_queue_bw bwspec,
+    struct node_queue_opt *opts)
+{
```
again too big.

Emmankoko committed:
fix err, return parens, sizeof args, and copyright issues

Emmankoko committed:
proper spacing between functions

zoulasc commented on sys/net/npf/npf_altq.c:
```diff
@@ -238,13 +238,13 @@ npftagname2tag(struct npf_tags *head, char *tagname)
+		return 0;
 	/* allocate and fill new struct npf_tagname */
-	tag = (struct npf_tagname *)malloc(sizeof(struct npf_tagname),
+	tag = (struct npf_tagname *)malloc(sizeof(*tag),
```
no cast in malloc

zoulasc commented on usr.sbin/npf/npfctl/npfctl_qstats.c:
```diff
@@ -146,18 +160,18 @@ npfctl_update_qstats(int fd, struct npf_altq_node **root)
-		err(1, "npfctl_insert_altq_node: calloc");
-	memcpy(&node->altq, &altq, sizeof(struct npf_altq));
+		err(EXIT_FAILURE, "npfctl_insert_altq_node: calloc");
+	memcpy(&node->altq, &altq, sizeof(altq));
```
it is safer to make it sizeof(dst) instead of sizeof(src)

Emmankoko replied:
right right right. but if there's a check, then we check if dst >= src (for overflow) right?

Emmankoko committed:
fix whitespaces

Emmankoko committed:
fix copyright

zoulasc commented on sys/altq/altq_priq.c:
```diff
@@ -39,6 +39,11 @@ __KERNEL_RCSID(0, "$NetBSD: altq_priq.c,v 1.28 2021/09/21 14:30:15 christos Exp
 
+#ifndef NPF
+#define NPF 1
+#endif
```
why is this needed?

zoulasc commented on sys/altq/altq_subr.c:
```diff
@@ -36,6 +36,11 @@ __KERNEL_RCSID(0, "$NetBSD: altq_subr.c,v 1.33 2017/03/14 09:03:08 ozaki-r Exp $
 #include "pf.h"
 #endif
 
+#ifndef NPF
```
Why is this needed?

zoulasc commented on sys/net/npf/npf.h:
```diff
@@ -309,6 +311,13 @@ typedef struct npf_ioctl_table {
 	} nct_data;
 } npf_ioctl_table_t;
 
+struct npfioc_altq {
```
unit32_t

Emmankoko committed:
Revert "Compile-time error fix in altq" and return 0

This reverts commit d9c78348f92fafe1a8b8735fc038cb2553c0a3ae.

Emmankoko committed:
fix whitespaces

Emmankoko committed:
use right copyright details

Emmankoko committed:
fix whitespace in scanner

Emmankoko committed:
fix whitespace in scanner

Emmankoko committed:
align bw spec whitespace

Emmankoko committed:
use sized types for queue opts

Emmankoko committed:
put everything in ALTQ defined block

Emmankoko committed:
fix whitespace in npftest Makefile

Emmankoko committed:
proper spacing in rule queue

Emmankoko committed:
full stop in copyright detail

Emmankoko committed:
clear verbosity

Emmankoko committed:
fix whitespace: put comments beside

Emmankoko committed:
rightly align fields

Emmankoko committed:
use right comment indennt

Emmankoko committed:
no cast in malloc, use sizeof(dest), and uintX_t

Emmankoko committed:
single responsibility principle: expand_altq

extract root queue creation and appending queue routine from expand_altq function to obtain a shorter funciton

Emmankoko committed:
quest for shorter functions: extract queue appending from expand queue

Emmankoko committed:
misc: fix missing return type on setqueue

Emmankoko committed:
add npf stuff to expected symbols

Emmankoko committed:
single responsibility principle: extract scheduler check from expand queue

Emmankoko committed:
fix sizeof and err in npfctl_qstats

Emmankoko committed:
terminating #ifndef in npf_altq.h

Emmankoko committed:
rename opts to use right objs

Emmankoko committed:
define opt in kernel option

Emmankoko committed:
terminate kernelndef

Emmankoko committed:
minor fix in npfaltq func prototype

Emmankoko committed:
provide def for npf_qid_unref

Emmankoko committed:
fix bad *

Emmankoko committed:
comment space

Emmankoko committed:
remove comment

Emmankoko committed:
use right open_dev sig

Emmankoko committed:
match build rule sig

Emmankoko committed:
fis minors spell and copyright

Emmankoko committed:
define struct segment

Emmankoko committed:
single responsibility: fix byte chackes

Emmankoko committed:
single responsibility: prototype check

Emmankoko committed:
add getaltq stuff

Emmankoko committed:
add getaltq stuff

Emmankoko committed:
properly check emptiness of ifname

Emmankoko committed:
avoid use of sizeof(*26675) in parser

Emmankoko committed:
add npf_altq to incs

Emmankoko committed:
Integrate: npf pseudo-device in altq

Emmankoko committed:
verbose check

Emmankoko committed:
add npf to kernel files

Emmankoko committed:
use right interface

Emmankoko committed:
use NPF_ALTQ flag into altq

Emmankoko committed:
check error

Emmankoko committed:
ifdef altq in ioctls

Emmankoko committed:
check compile

Emmankoko committed:
compiling

Emmankoko committed:
compiling

Emmankoko committed:
check error

Emmankoko committed:
needs-flags in npf kernel

Emmankoko committed:
needs-flags in npf kernel

Emmankoko committed:
add empty line to config

Emmankoko committed:
use right sig

Emmankoko committed:
verbose

Emmankoko committed:
verbose
